feat: :sparkles: comment dry run report

zolamk opened the pull request:
## Description

<!-- Please include a summary of the changes. -->
- send dry run report
<!-- Also include relevant motivation and context. -->
<!-- List any dependencies that are required for this change (if applicable.) -->

## Related issue

reviewpad/action#10

<!-- Closes # (issue) -->

## Type of change

<!-- Uncomment the right types of change from the options bellow: -->

<!-- Bug fix (non-breaking change which fixes an issue) -->
New feature (non-breaking change which adds functionality)
<!-- Improvements (non-breaking change without functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## How was this tested?

<!-- Please describe the tests that you ran to verify your changes. -->
<!-- Provide instructions so we can reproduce (if applicable.) -->
1. Run Reviewpad in dry run mode checked the results were as expected
<!-- Please also list any relevant details for your test configuration (if applicable.) -->

## Checklist

<!-- All checks are required in order to open a pull request ready to review. -->

- [x] I have performed a self-review of my code
- [x] I have ran `task check -f` and have no issues

reviewpad-team commented:
<!--@annotation-reviewpad-report-->
**Reviewpad Report**

**:information_source: Messages**
* @marcelosousa: you are being notified because critical code was modified

marcelosousa commented on lang/aladino/report.go:
```diff
@@ -65,21 +94,39 @@ func (report *Report) addToReport(statement *engine.Statement) {
 	// Header
-	sb.WriteString("**Reviewpad Report**\n\n")
+	if dryRun {
+		sb.WriteString("**Reviewpad Report** (Reviewpad ran in dry-run mode because reviewpad.yml has changed)\n\n")
```
We can actually pass a different input file. Instead of reviewpad.yml lets say that "because configuration file has changed"

marcelosousa commented on runner.go:
```diff
@@ -71,6 +71,14 @@ func Run(
 			engine.CollectError(evalEnv, err)
 			return nil, err
 		}
+	} else {
```
This can be problematic because we can run Reviewpad as a CLI in dry-run in which case the report will be added to the pull request. I think we need to introduce a new boolean flag to signal when the configuration file was changed.

zolamk replied:
should we also allow setting the new `reviewpadChanged` boolean from the cli?

zolamk replied:
Also, should it run in dry-run mode if the `reviewpadChanged` bool is true, or should we let the caller of the run function determine that?

marcelosousa replied:
I think the boolean variable should be called something on the lines `safeMode` instead of `reviewpadChanged`. 

I don't think there's need for the new boolean to be passed in the CLI for now. 

Regarding the second question, yes, it's a pre-condition that if the new boolean is true then dry-run needs to be true as well.

marcelosousa reviewed (changes requested):
Many thanks for this contribution 🙌 ! Please check the comment in runner.go which involves a substantial change in this pull request.

marcelosousa commented on lang/aladino/report.go:
```diff
@@ -42,6 +43,34 @@ func mergeReportWorkflowDetails(left, right ReportWorkflowDetails) ReportWorkflo
 }
 
+func (r *Report) SendReport(ctx context.Context, reviewpadFileChanged bool, mode string, pr *github.PullRequest, client *github.Client) error {
+
```
Please remove this empty line. Same in the rest of the pull request!

marcelosousa reviewed (changes requested):
Please check the style issue!

marcelosousa commented:
@zolamk can you please update with the base branch? The build is failing at the moment.

zolamk commented:
> @zolamk can you please update with the base branch? The build is failing at the moment.

I am rebasing at the moment

> @zolamk can you please update with the base branch? The build is failing at the moment.

I've rebased and updated the tests.

marcelosousa commented on runner.go:
```diff
@@ -43,7 +44,12 @@ func Run(
+	safeMode bool,
 ) (*engine.Program, error) {
+	if safeMode && !dryRun {
+		return nil, fmt.Errorf("reviewpad can only be run in dry-run when in safe mode")
```
The error message doesn't seem correct because you can run reviewpad with dry-run as true and safe mode as false. What you mean here is: "when reviewpad is running in safe mode, it must also run in dry-run".

marcelosousa reviewed (changes requested):
Check the comments in the runner.go. Thanks!

zolamk commented on runner.go:
```diff
@@ -73,6 +79,16 @@ func Run(
 	}
 
+	if safeMode {
+		report := aladino.ReportFromProgram(program)
+
+		err = report.SendReport(ctx, dryRun, reviewpadFile.Mode, pullRequest, client)
+		if err != nil {
+			engine.CollectError(evalEnv, err)
+			return nil, err
+		}
+	}
```
That was the first thing I tried too, but from what I can tell the aladino report is generated when the program is executed not during evaluation, which means in dry run mode since the program isn't executed there is no report generated and calling the `Report` method on the interpreter would just say `No workflows activated` [https://github.com/reviewpad/reviewpad/blob/62b1be11caa894e5cf04a252d72eae3cc807f29f/lang/aladino/interpreter.go#L131-L152](https://github.com/reviewpad/reviewpad/blob/62b1be11caa894e5cf04a252d72eae3cc807f29f/lang/aladino/interpreter.go#L131-L152)

zolamk replied:
What I see as a possible solution is during evaluation we could keep another report maybe called `EvaulationReport` and during `safeMode` we could have an `EvaluationReport` method on the interpreter and use that for the report. Sounds good?

marcelosousa replied:
I think it's confusing to have multiple reports. I think it's more elegant to pass the control variables to the interpreter environment and to use those variables inside the ExecProgram and ExecStatement interface method implementations. Does that sound good?

zolamk replied:
Yeah, that's a much better solution.

marcelosousa commented on runner.go:
```diff
@@ -59,18 +65,19 @@ func Run(
+		return nil, err
+	}
+
+	if safeMode {
```
This conditional is a no-op.

marcelosousa commented on lang/aladino/interpreter.go:
```diff
@@ -140,47 +145,31 @@ func (i *Interpreter) ExecStatement(statement *engine.Statement) error {
-	if comment == nil {
-		return AddReportComment(env, report)
-	}
+	return i.Env.GetReport().SendReport(env.GetCtx(), env.GetSafeMode(), mode, env.GetPullRequest(), env.GetClient())
```
This change increases the complexity of the code quite a bit. I think that abstracting the code with the `SendReport` makes things harder. I think we should revert that change.

Also, I think this is the only place where you really need to retrieve the `safeMode` boolean so I would prefer if you passed it as an argument to the `Report` function instead of adding it to the environment.

marcelosousa commented on runner.go:
```diff
@@ -59,18 +65,19 @@ func Run(
+		return nil, err
+	}
+
+	err = aladinoInterpreter.Report(reviewpadFile.Mode)
```
There's an issue with simply calling Report here. If we are running in safe mode then we want to add the report to the pull request, but if are not running in safe mode **and** in dry run, the comment will be added to the pull request and that is not intended.

marcelosousa commented on runner.go:
```diff
@@ -59,14 +65,16 @@ func Run(
+	}
 
-		err = aladinoInterpreter.Report(reviewpadFile.Mode)
+	// only send report if reviewpad is running in safe mode
+	// or not in dry run mode
```
this comment is not necessary :-)

marcelosousa commented on lang/aladino/interpreter.go:
```diff
@@ -170,17 +172,19 @@ func (i *Interpreter) Report(mode string) error {
 
 	if comment == nil {
-		return AddReportComment(env, report)
+		return AddReportComment(i.Env, report)
```
```suggestion
		return AddReportComment(env, report)
```

marcelosousa commented on lang/aladino/interpreter.go:
```diff
@@ -170,17 +172,19 @@ func (i *Interpreter) Report(mode string) error {
 	}
 
-	return UpdateReportComment(env, *comment.ID, report)
+	return UpdateReportComment(i.Env, *comment.ID, report)
```
```suggestion
	return UpdateReportComment(env, *comment.ID, report)
```

marcelosousa reviewed (commented):
I think this is the last round of comments before it is ready to merge! Thank you so much @zolamk 🚀

marcelosousa commented:
Many thanks for this contribution @zolamk 🚀
